### ISSUE TITLE
Slow and Ineffective training of LSTM(RNN) networks using Pytorch Estimator

### ISSUE THREAD

**Issue by GZHoffie** [intel-analytics/analytics-zoo#3557]
I was trying to use a LSTM network to train a text predictor on the [nietsche](https://s3.amazonaws.com/text-datasets/nietzsche.txt) dataset. My network looks like the following
```python
import torch.nn as nn
import torch.nn.functional as F

class LSTMNet(nn.Module):
    def __init__(self, maxlen, chars):
        super(LSTMNet, self).__init__()
        self.embedding = nn.Embedding(len(chars), maxlen)
        self.lstm = nn.LSTM(maxlen, 128)
        self.fc = nn.Linear(128, len(chars))
        
        self.maxlen = maxlen
        self.chars = chars

    def forward(self, x):
        batch_size, vocab_size = x.size()
        x = self.embedding(x)
        h0 = torch.randn(1, vocab_size, 128)
        c0 = torch.randn(1, vocab_size, 128)
        (x, _) = self.lstm(x, (h0, c0))
        return F.log_softmax(self.fc(x[:, -1, :]), dim=-1)
        
model = LSTMNet(maxlen, chars)
model.train()
criterion = nn.CrossEntropyLoss()
opt = torch.optim.RMSprop(model.parameters(), lr=0.01)
```
which is just an embedding layer LSTM network followed by a fully-connected layer. This network is a copy of the [Keras version](https://github.com/GZHoffie/zoo-tutorials/blob/master/orca/8.1-text-generation-with-lstm.ipynb) and the only difference (I think) is the embedding layer.

However the network appears to be inefficient:
* It is very slow while training (with throughput around 1,000 data / second), leading to about 200 seconds per epoch, much slower than the Keras version. The speed becomes even slower to around 600 data / second in later epochs (~50 epochs).
* The loss doesn't decrease much after the first epoch (converges around 2.5 while starting at around 4.0), and the predictions given by the network are not as good as the Keras network.

I tried to adjust the learning rate, the batch size, and also the initialization of `h0` and `c0`, and tried storing `h0` and `c0` using the following:
```python
    def __init__(self, maxlen, chars):
        ...
        self.h0 = torch.randn(1, vocab_size, 128)
        self.c0 = torch.randn(1, vocab_size, 128)

    def forward(self, x):
        ...
        (x, (self.h0, self.c0)) = self.lstm(x, (self.h0, self.c0))
        self.h0 = self.h0.detach()
        self.c0 = self.c0.detach()
        ...
```
But none of them worked. The rest to check is the `DataLoader` and if the network is built correctly.
For the full code please see this [notebook](https://github.com/GZHoffie/zoo-tutorials/blob/pytorch8.1/pytorch/8.1-text-generation-with-lstm.ipynb). Thank you!
